botocove/cove_host_account.py

```diff
@@ -81,175 +94,229 @@ def get_cove_sessions(self) -> List[CoveSessionInformation]:
+        accounts.update(resolver.resolve_accounts(res))
+    return accounts
 
-        return account_list
 
-    def _get_all_child_ous(self, parent_ou: str, ou_list: List[str]) -> None:
-        """Depth-first recursion mutates the current_ou_list present in the calling
-        function to establish all children of a parent OU"""
-        child_ous = self._get_child_ous(parent_ou)
-        child_ous_list = [ou["Id"] for ou in child_ous["Children"]]
-        ou_list.extend(child_ous_list)
+class Resolver(Protocol):
+    def resolve_accounts(self, resource_id: str) -> Set[str]:
+        ...
 
-        for ou in child_ous_list:
-            self._get_all_child_ous(ou, ou_list)
 
-    def _get_accounts_by_organization_units(
-        self, organization_units: List[str]
-    ) -> List[str]:
+class AccountResolver:
+    def resolve_accounts(self, resource_id: str) -> Set[str]:
+        if _is_account_id(resource_id):
+            return {resource_id}
+        raise ValueError(f"provided id is not an aws account: {resource_id}")
 
-        account_list: List[str] = []
 
-        for ou in organization_units:
-            ou_children = self._get_child_accounts(ou)
-            account_list.extend(acc["Id"] for acc in ou_children["Children"])
+class OrganizationalUnitResolver:
+    def __init__(self, org_client: OrganizationsClient) -> None:
+        self._org_client = org_client
 
-        return account_list
+    def resolve_accounts(self, resource_id: str) -> Set[str]:
+        if _is_organizational_unit_id(resource_id):
+            return _list_descendant_accounts(self._org_client, resource_id)
+        raise ValueError(f"provided id is not an ou: {resource_id}")
 
-    def _get_active_org_accounts(self) -> Set[str]:
-        all_org_accounts = (
-            self.org_client.get_paginator("list_accounts")
-            .paginate()
-            .build_full_result()["Accounts"]
-        )
-        return {acc["Id"] for acc in all_org_accounts if acc["Status"] == "ACTIVE"}
-
-    @lru_cache()
-    def _get_child_ous(self, parent_ou: str) -> ListChildrenResponseTypeDef:
-        return cast(
-            ListChildrenResponseTypeDef,
-            self.org_client.get_paginator("list_children")
-            .paginate(ChildType="ORGANIZATIONAL_UNIT", ParentId=parent_ou)
-            .build_full_result(),
+
+class AccountAndOrganizationalUnitResolver:
+    def __init__(self, org_client: OrganizationsClient) -> None:
+        self._org_client = org_client
+
+    def resolve_accounts(self, resource_id: str) -> Set[str]:
+        resolvers: List[Resolver] = [
+            AccountResolver(),
+            OrganizationalUnitResolver(self._org_client),
+        ]
+        for resolver in resolvers:
+            try:
+                return resolver.resolve_accounts(resource_id)
+            except ValueError:
+                continue
+        raise ValueError(
+            f"provided id is neither an aws account nor an ou: {resource_id}"
```

Comment:
Think this might have been mangled somewhere? It looks like line 206 is unreachable and `AccountResolver()`  is the only resolver that'll run?

Reply:
The older `test_malformed_ignore_ids` checks the the "provided id" ValueError is raised for a nonsense value `cat`. If it passes then this line must be reachable because it's the only part that generates the error.

The general pattern here is to return from first resolver that doesn't raise its own ValueError exception. If AccountResolver raises ValueError (because the resource ID is not an account ID) then we try OrganizationalUnitResolver. If that also raises ValueError (because the resource ID is not an organizational unit ID) then there are no more resolvers to try and the final ValueError is raised.

It's a bit convoluted but it was the cleanest thing I could think of to maintain the existing behavior.